californium-core/src/main/java/org/eclipse/californium/core/network/UdpMatcher.java

```diff
@@ -310,11 +317,19 @@ public void completed(final Exchange exchange) {
 				} else {
 					// in case an empty ACK was lost
 					KeyMID idByMID = KeyMID.fromOutboundMessage(originRequest);
-					exchangeStore.remove(idByMID);
+					exchangeStore.remove(idByMID, exchange);
 				}
 				KeyToken idByToken = KeyToken.fromOutboundMessage(originRequest);
-				exchangeStore.remove(idByToken);
-				if(!originRequest.getOptions().hasObserve()) {
+				exchangeStore.remove(idByToken, exchange);
+				Request request = exchange.getRequest();
+				if (request != originRequest && null != request &&  null != request.getToken()
+						&& !Arrays.equals(request.getToken(), originRequest.getToken())) {
+					// remove starting request also
+					originRequest = request;
+					idByToken = KeyToken.fromOutboundMessage(originRequest);
+					exchangeStore.remove(idByToken, exchange);
```

Comment:
Ok but I believe that the idea about the blocklayer refactoring was to move as much as possible block2 behavior to this class. I also believe that we all would like to removing this confusing currentRequest/request. It seems a good strategy could be to isolate it before to remove it.